File: trunk/Source/Runtime/CPlusplus/Animation/Bibim/EvalUnitRandom.cpp
#include <Bibim/PCH.h>
#include <Bibim/EvalUnitRandom.h>
#include <Bibim/Clock.h>
#include <Bibim/ComponentStreamReader.h>
#include <Bibim/Math.h>

namespace Bibim
{
    BBImplementsComponent(EvalUnitRandom);

    EvalUnitRandom::EvalUnitRandom()
        : value(0.0f),
          seed(0),
          initialSeed(0),
          cycle(EachTime)
    {
    }

    EvalUnitRandom::~EvalUnitRandom()
    {
    }

    void EvalUnitRandom::Generate()
    {
        const int generatedValue = seed % 1001;
        value = static_cast<float>(generatedValue) * 0.001f;
        seed = 49327 * (seed & 0xABCDABCD) + (seed >> 16);

        if (seed == 0)
        {
            if (initialSeed != 0)
                seed = initialSeed;
            else
                seed = static_cast<int>(Clock::GetCurrentMilliSeconds());
        }
    }

    void EvalUnitRandom::Reset()
    {
        if (cycle == ResetOrManual)
            Generate();
    }

    float EvalUnitRandom::Evaluate(EvalContext& /*context*/)
    {
        if (cycle == EachTime)
            Generate();

        return value;
    }

    void EvalUnitRandom::SetInitialSeed(int value)
    {
        value = Math::Max(value, 0);
        if (initialSeed != value)
        {
            initialSeed = value;

            if (initialSeed != 0)
                seed = initialSeed;
            else
                seed = static_cast<int>(Clock::GetCurrentMilliSeconds());

            Generate();
        }
    }

    void EvalUnitRandom::OnRead(ComponentStreamReader& reader)
    {
        Base::OnRead(reader);
        initialSeed = reader.ReadInt();
        cycle = static_cast<GenerationCycle>(reader.ReadByte());
        seed = initialSeed;
        Generate();
    }

    void EvalUnitRandom::OnCopy(const GameComponent* original, CloningContext& context)
    {
        Base::OnCopy(original, context);
        const This* o = static_cast<const This*>(original);
        value = o->value;
        seed = o->seed;
        initialSeed = o->initialSeed;
        cycle = o->cycle;
    }
}

File: trunk/Source/Runtime/CSharp/Bibim.Runtime/Animation/Evaluation/EvalUnitRandom.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Bibim.Animation
{
    [ClassID('e', 'U', 'R', 'f')]
    public sealed class EvalUnitRandom : Eval<float>
    {
        #region GenerationCycle (Nested Enum)
        public enum GenerationCycle
        {
            FullManual,
            ResetOrManual,
            EachTime,
        }
        #endregion

        #region Fields
        private float value;
        private int initialSeed;
        private Random random;
        #endregion

        #region Properties
        public int InitialSeed
        {
            get { return initialSeed; }
            set
            {
                value = Math.Max(value, 0);
                if (initialSeed != value)
                {
                    initialSeed = value;
                    if (initialSeed != 0)
                        random = new Random(initialSeed);
                    else
                        random = new Random();
                    Generate();
                }
            }
        }

        public GenerationCycle Cycle
        {
            get;
            set;
        }
        #endregion

        #region Constructors
        public EvalUnitRandom()
            : this(GenerationCycle.EachTime)
        {
        }

        public EvalUnitRandom(GenerationCycle cycle)
        {
            value = 0.0f;
            Cycle = cycle;
            initialSeed = 0;
            random = new Random();
            Generate();
        }
        #endregion

        public void Generate()
        {
            int generatedValue = random.Next(10000);
            value = ((float)generatedValue) * 0.0001f;
        }

        public override void Start()
        {
            Reset();
        }

        public override void Stop()
        {
        }

        public override void Reset()
        {
            if (Cycle == GenerationCycle.ResetOrManual)
                Generate();
        }

        public override float Evaluate(EvalContext context)
        {
            if (Cycle == GenerationCycle.EachTime)
                Generate();

            return value;
        }
    }
}
